--- a/qtwebkit-examples-and-demos/examples/webkitwidgets/browser/data/defaultbookmarks.xbel	2012-12-18 14:04:47.000000000 -0500
+++ b/qtwebkit-examples-and-demos/examples/webkitwidgets/browser/data/defaultbookmarks.xbel	2012-12-30 12:56:20.837594558 -0500
@@ -4,34 +4,25 @@
     <folder folded="yes">
         <title>Bookmarks Bar</title>
         <bookmark href="http://qt-project.org/">
-            <title>Qt Home Page</title>
+            <title>Qt Home</title>
         </bookmark>
         <bookmark href="http://webkit.org/">
             <title>WebKit.org</title>
         </bookmark>
         <bookmark href="http://qt-project.org/doc/">
-            <title>Qt Documentation</title>
-        </bookmark>
-        <bookmark href="http://qt-project.org/quarterly/">
-            <title>Qt Quarterly</title>
-        </bookmark>
-        <bookmark href="http://planet.qt-project.org/">
-            <title>Qt Blog</title>
-        </bookmark>
-        <bookmark href="http://www.qtcentre.org/">
-            <title>Qt Centre</title>
+            <title>Qt Docs</title>
         </bookmark>
-        <bookmark href="http://qt-apps.org/">
-            <title>Qt-Apps.org</title>
+        <bookmark href="http://chakra-project.org/index.html">
+            <title>Chakra-Project</title>
         </bookmark>
-        <bookmark href="http://qt-project.org/wiki/OnlineCommunities/">
-            <title>Online Communities</title>
+        <bookmark href="http://chakra-project.org/bbs/">
+            <title>Forum</title>
         </bookmark>
-       <bookmark href="http://xkcd.com/">
-            <title>xkcd</title>
+        <bookmark href="http://webchat.freenode.net/?channels=chakra">
+            <title>IRC</title>
         </bookmark>
-       <bookmark href="http://twitter.com/qtproject">
-            <title>Twitter</title>
+        <bookmark href="http://www.youtube.com/html5">
+            <title>YouTube/html5</title>
         </bookmark>
     </folder>
     <folder folded="yes">
