--- a/modules/codec/dav1d.c	2022-02-21 10:19:25.000000000 +0100
+++ b/modules/codec/dav1d.c	2022-03-21 13:10:46.215486199 +0100
@@ -63,10 +63,9 @@
     set_category(CAT_INPUT)
     set_subcategory(SUBCAT_INPUT_VCODEC)
 
-    add_integer_with_range("dav1d-thread-frames", 0, 0, DAV1D_MAX_FRAME_THREADS,
-                THREAD_FRAMES_TEXT, THREAD_FRAMES_LONGTEXT, false)
-    add_integer_with_range("dav1d-thread-tiles", 0, 0, DAV1D_MAX_TILE_THREADS,
-                THREAD_TILES_TEXT, THREAD_TILES_LONGTEXT, false)
+    add_integer_with_range("dav1d-thread-frames", 0, 0, DAV1D_MAX_THREADS,
+                THREAD_FRAMES_TEXT, THREAD_FRAMES_LONGTEXT, false)
+    add_obsolete_string("dav1d-thread-tiles") // unused with dav1d 1.0
 vlc_module_end ()
 
 /*****************************************************************************
@@ -294,15 +293,9 @@
         return VLC_ENOMEM;
 
     dav1d_default_settings(&p_sys->s);
-    p_sys->s.n_tile_threads = var_InheritInteger(p_this, "dav1d-thread-tiles");
-    if (p_sys->s.n_tile_threads == 0)
-        p_sys->s.n_tile_threads =
-            (i_core_count > 4) ? 4 :
-            (i_core_count > 1) ? i_core_count :
-            1;
-    p_sys->s.n_frame_threads = var_InheritInteger(p_this, "dav1d-thread-frames");
-    if (p_sys->s.n_frame_threads == 0)
-        p_sys->s.n_frame_threads = (i_core_count < 16) ? i_core_count : 16;
+    p_sys->s.n_threads = var_InheritInteger(p_this, "dav1d-thread-frames");
+    if (p_sys->s.n_threads == 0)
+        p_sys->s.n_threads = __MAX(1, vlc_GetCPUCount());
     p_sys->s.allocator.cookie = dec;
     p_sys->s.allocator.alloc_picture_callback = NewPicture;
     p_sys->s.allocator.release_picture_callback = FreePicture;
@@ -313,12 +306,12 @@
         return VLC_EGENERIC;
     }
 
-    msg_Dbg(p_this, "Using dav1d version %s with %d/%d frame/tile threads",
-            dav1d_version(), p_sys->s.n_frame_threads, p_sys->s.n_tile_threads);
+    msg_Dbg(p_this, "Using dav1d version %s with %d threads",
+            dav1d_version(), p_sys->s.n_threads);
 
     dec->pf_decode = Decode;
     dec->pf_flush = FlushDecoder;
-    dec->i_extra_picture_buffers = (p_sys->s.n_frame_threads - 1);
+    dec->i_extra_picture_buffers = (p_sys->s.n_threads - 1);
 
     dec->fmt_out.video.i_width = dec->fmt_in.video.i_width;
     dec->fmt_out.video.i_height = dec->fmt_in.video.i_height;
