diff -Naur vlc-3.0.3/modules/gui/qt/components/simple_preferences.cpp vlc-3.0.3-p/modules/gui/qt/components/simple_preferences.cpp
--- vlc-3.0.3/modules/gui/qt/components/simple_preferences.cpp	2018-03-07 14:52:20.000000000 +0100
+++ vlc-3.0.3-p/modules/gui/qt/components/simple_preferences.cpp	2018-05-29 00:07:18.708291085 +0200
@@ -45,6 +45,7 @@
 #include <QScreen>
 #include <QtAlgorithms>
 #include <QDir>
+#include <QButtonGroup>
 #include <assert.h>
 #include <math.h>
