--- a/libmirage/CMakeLists.txt	2013-06-07 15:34:16.000000000 -0400
+++ b/libmirage/CMakeLists.txt	2013-06-07 21:41:52.201637073 -0400
@@ -31,7 +31,7 @@
 include (FileList)
 
 # Options
-option (GTKDOC_ENABLED "Use gtk-doc to build documentation" on)
+option (GTKDOC_ENABLED "Use gtk-doc to build documentation" off)
 option (INTROSPECTION_ENABLED "Generate gobject introspection files" on)
 option (POST_INSTALL_HOOKS "Run post-install hooks" on)
 
