diff --git a/download.c b/download.c
index 64dabe5..98312b7 100644
--- a/download.c
+++ b/download.c
@@ -71,7 +71,7 @@ clamz_downloader *new_downloader(const clamz_config *cfg)
     curl_easy_setopt(dl->curl, CURLOPT_FOLLOWLOCATION, 1L);
     curl_easy_setopt(dl->curl, CURLOPT_COOKIEFILE, "");
 
-    sprintf(useragent, "Amazon MP3 Downloader (%s)", PACKAGE_STRING);
+    sprintf(useragent, "Amazon MP3 Downloader (Win32 1.0.17 en_US)");
     curl_easy_setopt(dl->curl, CURLOPT_USERAGENT, useragent);
 
     cookiejar = get_config_file_name(NULL, "cookies", NULL);
