--- src/core/config/desktop_linux.pri.orig	2014-04-09 23:32:42.144599016 -0400
+++ src/core/config/desktop_linux.pri	2014-04-09 23:32:42.471274036 -0400
@@ -3,16 +3,16 @@
 GYP_CONFIG += \
     desktop_linux=1 \
     toolkit_uses_gtk=0 \
-    use_alsa=0 \
+    use_alsa=1 \
     use_aura=1 \
     use_ash=0 \
-    use_cairo=0 \
+    use_cairo=1 \
     use_clipboard_aurax11=0 \
-    use_cups=0 \
+    use_cups=1 \
     use_gconf=0 \
     use_gio=0 \
     use_gnome_keyring=0 \
     use_kerberos=0 \
-    use_pango=0 \
-    use_pulseaudio=0 \
+    use_pango=1 \
+    use_pulseaudio=1 \
 
