--- a/src/3rdparty/chromium/third_party/blink/renderer/build/scripts/rule_bison.py	2020-09-27 22:53:46.441526029 +0200
+++ b/src/3rdparty/chromium/third_party/blink/renderer/build/scripts/rule_bison.py	2020-09-27 22:53:46.361526021 +0200
@@ -46,8 +46,8 @@
 from blinkbuild.name_style_converter import NameStyleConverter
 
 def modify_file(path, prefix_lines, suffix_lines, replace_list=[]):
-    prefix_lines = map(lambda s: s + '\n', prefix_lines)
-    suffix_lines = map(lambda s: s + '\n', suffix_lines)
+    prefix_lines = list(map(lambda s: s + '\n', prefix_lines))
+    suffix_lines = list(map(lambda s: s + '\n', suffix_lines))
     with open(path, 'r') as f:
         old_lines = f.readlines()
     for i in range(len(old_lines)):
