--- dbus-sharp-0.7.0/Makefile.in	2010-09-14 11:18:28.000000000 +0200
+++ dbus-sharp-0.7.0/Makefile.in	2011-09-04 01:19:56.000000000 +0200
@@ -203,7 +203,7 @@ target_alias = @target_alias@
 top_build_prefix = @top_build_prefix@
 top_builddir = @top_builddir@
 top_srcdir = @top_srcdir@
-SUBDIRS = src tools examples
+SUBDIRS = src tools
 pkgconfigdir = $(libdir)/pkgconfig
 pkgconfig_DATA = dbus-sharp-1.0.pc
 EXTRA_DIST = \
---  dbus-sharp-0.7.0/Makefile.am	2010-09-14 11:17:49.000000000 +0200
+++  dbus-sharp-0.7.0/Makefile.am.nogtk	2011-09-04 01:18:46.000000000 +0200
@@ -1,4 +1,4 @@
-SUBDIRS = src tools examples
+SUBDIRS = src tools
 
 pkgconfigdir = $(libdir)/pkgconfig
 pkgconfig_DATA = dbus-sharp-1.0.pc
