--- larch.orig/cli/medium.py  2011-03-20 09:43:11.725196269 +0100
+++ larch/cli/medium.py   2011-03-20 09:43:11.741862869 +0100
@@ -99,7 +99,7 @@ def build_medium(options, device):
                     modsdst = build
                     runcomd('mkdir -p %s/larch' % modsdst)
                 if not runcmd('%s/boot/support/support mksquashfs %s %s/larch/mods.sqf'
-                        ' -wildcards -e %s'
+                        ' -b 256K -Xbcj x86 -wildcards -e %s'
                         % (build, medium_dir, modsdst, IGNOREDIRS),
                         filter=mksquashfs_filter_gen())[0]:
                     errout(_("Squashing mods.sqf failed"))
--- larch.orig/cli/larchify.py        2011-03-20 09:42:37.658662461 +0100
+++ larch/cli/larchify.py 2011-03-20 09:42:37.678662382 +0100
@@ -128,7 +128,7 @@ class Builder:
                 fh.close()
 
             if not chroot(self.installation0,
-                    "/sbin/mksquashfs '/' '%s' -wildcards -e %s"
+                    "/sbin/mksquashfs '/' '%s' -b 256K -Xbcj x86 -wildcards -e %s"
                     % (CHROOT_SYSTEMSQF, ignoredirs),
                     filter=mksquashfs_filter_gen()):
                 errout(_("Squashing system.sqf failed"))
@@ -253,7 +253,7 @@ class Builder:
 
         comment("Squashing mods.sqf")
         if not chroot(self.installation0,
-                "/sbin/mksquashfs '%s' '%s/larch/mods.sqf' -wildcards -e %s"
+                "/sbin/mksquashfs '%s' '%s/larch/mods.sqf' -b 256K -Xbcj x86 -wildcards -e %s"
                 % (CHROOT_DIR_OVERLAY, CHROOT_DIR_MEDIUM, IGNOREDIRS),
                 filter=mksquashfs_filter_gen()):
             errout(_("Squashing mods.sqf failed"))
