diff --git a/larch/PKGBUILD b/larch/PKGBUILD
index 177c7c9..b9c5e21 100644
--- a/larch/PKGBUILD
+++ b/larch/PKGBUILD
@@ -2,7 +2,7 @@
 # 2011.01.30
 pkgname=larch
 pkgver=8.2.17
-pkgrel=4
+pkgrel=5
 
 pkgdesc="Design your own 'live' Chakra GNU/Linux DVDs, CDs and USB-sticks (etc.)"
 url="http://larch.berlios.de"
diff --git a/larch/cli/config.py b/larch/cli/config.py
index b7f0c37..258795e 100644
--- a/larch/cli/config.py
+++ b/larch/cli/config.py
@@ -52,7 +52,7 @@ PACMAN_CONF = '/tmp/larch_pacman.conf'
 # Medium detection alternatives
 detection_methods = 'label|uuid|device|search'
 # Directories to ignore when squashing mods.sqf
-IGNOREDIRS = 'boot dev larch media proc sys tmp .*'
+IGNOREDIRS = 'dev larch media proc sys tmp .*'
 # Valid file-system types for extlinux
 OKFS = ('ext2', 'ext3', 'ext4', 'btrfs')
 
diff --git a/larch/cli/larchify.py b/larch/cli/larchify.py
index 3b039f9..30ed14b 100755
--- a/larch/cli/larchify.py
+++ b/larch/cli/larchify.py
@@ -110,6 +110,17 @@ class Builder:
         writefile(self.kname, self.medium_dir + '/boot/kernelname')
         if os.path.isfile("%s/boot/%s" % (self.installation0, 'vmlinuz26-lts')):
             writefile('vmlinuz26-lts', self.medium_dir + '/boot/kernelname-lts')
+            
+        # if burg folder found we backup it and restore it
+        if os.path.isfile("%s/boot/burg" % (self.installation0)):
+            comment("Remove any /boot folder but restore found burg folder")
+            runcmd("cp -Rv %s/boot/burg %s/tmp" % (self.installation0, self.installation0))
+            runcmd("rm -Rfv %s/boot" % (self.installation0))
+            runcmd("mkdir -p %s/boot" % (self.installation0))
+            runcmd("cp -Rv %s/tmp/burg %s/boot" % (self.installation0, self.installation0))
+        else:
+            comment("Remove any /boot folder")
+            runcmd("rm -Rfv %s/boot" % (self.installation0))
 
         # if no saved system.sqf, squash the Arch installation at self.installation_dir
         if not os.path.isfile(self.system_sqf):
