--- a/filter/pdftoopvp/OPVPOutputDev.cxx	2015-06-26 11:45:26.000000000 -0400
+++ b/filter/pdftoopvp/OPVPOutputDev.cxx	2015-07-26 17:20:55.761900689 -0400
@@ -1812,7 +1812,7 @@
   maskSplash->drawImage(&imageSrc, &imgMaskData,
 			splashModeMono8, gFalse, maskWidth, maskHeight, mat);
 #else
-  maskSplash->drawImage(&imageSrc, &imgMaskData,
+  maskSplash->drawImage(&imageSrc, NULL, &imgMaskData,
 			splashModeMono8, gFalse, maskWidth, maskHeight,
                         mat,gFalse);
 #endif
--- a/filter/pdftoopvp/oprs/OPRS.cxx	2015-06-26 11:45:26.000000000 -0400
+++ b/filter/pdftoopvp/oprs/OPRS.cxx	2015-07-26 17:26:25.810644127 -0400
@@ -241,7 +241,7 @@
 #if POPPLER_VERSION_MAJOR <= 0 && (POPPLER_VERSION_MINOR <= 20 || (POPPLER_VERSION_MINOR == 21 && POPPLER_VERSION_MICRO <= 2))
 	return splash->drawImage(src,srcData,srcMode,srcAlpha,w,h,mat);
 #else
-	return splash->drawImage(src,srcData,srcMode,srcAlpha,w,h,mat,gFalse);
+	return splash->drawImage(src,NULL,srcData,srcMode,srcAlpha,w,h,mat,gFalse);
 #endif
     } else {
 	return opvpSplash->drawImage(src,srcData,srcMode,srcAlpha,w,h,mat);
