--- gfxboot-4.2.2-orig/gfxboot	2010-04-16 15:19:43.000000000 +0000
+++ gfxboot-4.2.2/gfxboot	2012-07-27 15:12:27.000000000 +0000
@@ -986,9 +986,9 @@
 my %bl_list = (
   grub     => '/usr/sbin/grub',
   lilo     => '/sbin/lilo',
-  isolinux => '/usr/share/syslinux/isolinux.bin',
+  isolinux => '/usr/lib/syslinux/isolinux.bin',
   syslinux => '/usr/bin/syslinux',
-  pxelinux => '/usr/share/syslinux/pxelinux.0',
+  pxelinux => '/usr/lib/syslinux/pxelinux.0',
   bd       => '/usr/bin/bd',
   bochs    => '/usr/bin/bochs',
 );
@@ -1782,8 +1782,8 @@
   $arch_dir = 'i386';
   $arch_dir = 'x86_64' if $opt_64 && !$opt_32;
 
-  $comboot = "$opt_syslinux/usr/share/syslinux/gfxboot.c32";
-  $comboot = "$opt_syslinux/usr/share/syslinux/gfxboot.com" unless -f $comboot;
+  $comboot = "$opt_syslinux/usr/lib/syslinux/gfxboot.c32";
+  $comboot = "$opt_syslinux/usr/lib/syslinux/gfxboot.com" unless -f $comboot;
   $comboot = 0 unless -f $comboot;
 
   my $menu = fake_menu 'install';
@@ -1853,7 +1853,7 @@
     close F;
   }
 
-  system "cp $opt_syslinux/usr/share/syslinux/isolinux.bin $dst/$loader" and die "error: no isolinux\n";
+  system "cp $opt_syslinux/usr/lib/syslinux/isolinux.bin $dst/$loader" and die "error: no isolinux\n";
   system "cp $comboot $dst/$loader" if $comboot;
 
   for (@opt_test_addfiles) {
@@ -1897,8 +1897,8 @@
 
   die "error: syslinux not found\n" unless -f "$opt_syslinux/$bl_list{syslinux}";
 
-  $comboot = "$opt_syslinux/usr/share/syslinux/gfxboot.c32";
-  $comboot = "$opt_syslinux/usr/share/syslinux/gfxboot.com" unless -f $comboot;
+  $comboot = "$opt_syslinux/usr/lib/syslinux/gfxboot.c32";
+  $comboot = "$opt_syslinux/usr/lib/syslinux/gfxboot.com" unless -f $comboot;
   $comboot = 0 unless -f $comboot;
 
   my $menu = fake_menu 'install';
@@ -2011,8 +2011,8 @@
   $arch_dir = 'i386';
   $arch_dir = 'x86_64' if $opt_64 && !$opt_32;
 
-  $comboot = "$opt_syslinux/usr/share/syslinux/gfxboot.c32";
-  $comboot = "$opt_syslinux/usr/share/syslinux/gfxboot.com" unless -f $comboot;
+  $comboot = "$opt_syslinux/usr/lib/syslinux/gfxboot.c32";
+  $comboot = "$opt_syslinux/usr/lib/syslinux/gfxboot.com" unless -f $comboot;
   $comboot = 0 unless -f $comboot;
 
   my $menu = fake_menu 'install';
@@ -2082,7 +2082,7 @@
     close F;
   }
 
-  system "cp $opt_syslinux/usr/share/syslinux/pxelinux.0 $dst/$loader" and die "error: no pxelinux\n";
+  system "cp $opt_syslinux/usr/lib/syslinux/pxelinux.0 $dst/$loader" and die "error: no pxelinux\n";
   system "cp $comboot $dst/$loader" if $comboot;
 
   for (@opt_test_addfiles) {
   