--- a/kernel/nvidia-drm/nvidia-dma-resv-helper.h
+++ b/kernel/nvidia-drm/nvidia-dma-resv-helper.h
@@ -39,6 +39,15 @@
 #include <linux/reservation.h>
 #endif
 
+#include <linux/version.h>
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(5, 19, 0))
+// Rel. commit "dma-buf: specify usage while adding fences to dma_resv obj v7" (Christian König, 9 Nov 2021)
+static inline void dma_resv_add_excl_fence(struct dma_resv *obj, struct dma_fence *fence)
+{
+    dma_resv_add_fence(obj, fence, DMA_RESV_USAGE_WRITE);
+}
+#endif
+
 #include <nvidia-dma-fence-helper.h>
 
 #if defined(NV_LINUX_DMA_RESV_H_PRESENT)
