--- a/configure	2015-07-22 11:33:56.137867946 -0700
+++ b/configure	2015-07-22 11:32:23.208465360 -0700
@@ -20168,7 +20168,7 @@
 
 
 	# Extract the first word of "gmcs", so it can be a program name with args.
-set dummy gmcs; ac_word=$2
+set dummy mcs; ac_word=$2
 { $as_echo "$as_me:${as_lineno-$LINENO}: checking for $ac_word" >&5
 $as_echo_n "checking for $ac_word... " >&6; }
 if ${ac_cv_path_MCS+:} false; then :

