--- a/m4/flex.m4
+++ b/m4/flex.m4
@@ -8,11 +8,11 @@ dnl ad_AC_PROG_FLEX
 dnl Look for flex or missing, then run AC_PROG_LEX and AC_DECL_YYTEXT
 AC_DEFUN(ad_AC_PROG_FLEX,
 [AC_CHECK_PROGS(LEX, flex, missing)
-if test "$LEX" = missing; then
+AS_IF([test "$LEX" = missing], [dnl
   LEX="\$(top_srcdir)/$ac_aux_dir/missing flex"
   LEX_OUTPUT_ROOT=lex.yy
   AC_SUBST(LEX_OUTPUT_ROOT)dnl
-else
+], [:
   AC_PROG_LEX
   AC_DECL_YYTEXT
-fi])
+])])

--- recode-3.6/m4/gettext.m4
+++ recode-3.6/m4/gettext.m4
@@ -109,12 +109,12 @@
     else
       ac_items="$LINGUAS"
       for ac_item in $ac_items; do
-	case "$ALL_LINGUAS" in
-	  *$ac_item*)
+        for supported_item in $ALL_LINGUAS; do
+          if test "$ac_item" = "$supported_item"; then
 	    ac_print="$ac_print $ac_item"
 	    MOFILES="$MOFILES $ac_item.mo"
-	    ;;
-	esac
+          fi
+        done
       done
     fi
     AC_SUBST(MOFILES)
     


