--- configure.ac	2012-06-03 19:32:35.019700768 +0200
+++ configure.ac.new	2012-06-03 19:38:02.479753466 +0200
@@ -62,7 +62,7 @@
 GTK_DOC_CHECK
 
 dnl Check for Glib 
-PKG_CHECK_MODULES(GLIB, gobject-2.0 >= 2.30, have_glib=yes, have_glib=no)
+PKG_CHECK_MODULES(GLIB, [gobject-2.0 >= 2.30, gio-2.0 >= 2.30], have_glib=yes, have_glib=no)
 
 if test x$have_glib = xno ; then
     AC_MSG_ERROR([GLib development libraries not found])
