--- a/glm/simd/platform.h	2017-08-16 16:48:40.000000000 +0200
+++ b/glm/simd/platform.h	2018-04-22 14:20:21.247939536 +0200
@@ -281,7 +281,7 @@
 #		define GLM_COMPILER (GLM_COMPILER_GCC70)
 #	elif (__GNUC__ == 7) && (__GNUC_MINOR__ == 1)
 #		define GLM_COMPILER (GLM_COMPILER_GCC71)
-#	elif (__GNUC__ == 7) && (__GNUC_MINOR__ == 2)
+#	elif (__GNUC__ == 7) && (__GNUC_MINOR__ == 3)
 #		define GLM_COMPILER (GLM_COMPILER_GCC72)
 #	elif (__GNUC__ >= 8)
 #		define GLM_COMPILER (GLM_COMPILER_GCC80)
