From bda961062fa784647ea8d1d2e97b5abb9297ef19 Mon Sep 17 00:00:00 2001
From: Timo Gurr <timo.gurr@gmail.com>
Date: Mon, 26 Feb 2024 18:11:58 +0100
Subject: [PATCH] build: do not build auth-dialog when gnome support is
 disabled

---
 Makefile.am | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/Makefile.am b/Makefile.am
index 17ec0e6..9455614 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -292,6 +292,7 @@ properties_tests_test_import_export_glib_LDADD = \
 
 ###############################################################################
 
+if WITH_GNOME
 libexec_PROGRAMS += auth-dialog/nm-vpnc-auth-dialog
 
 auth_dialog_nm_vpnc_auth_dialog_CPPFLAGS = \
@@ -308,6 +309,7 @@ auth_dialog_nm_vpnc_auth_dialog_LDADD = \
 	$(LIBNM_LIBS) \
 	$(LIBNMA_LIBS) \
 	$(LIBSECRET_LIBS)
+endif
 
 ###############################################################################
 
-- 
2.44.0


