diff --git a/texk/web2c/pdftexdir/utils.c b/texk/web2c/pdftexdir/utils.c
index c93a8781..6f866e76 100644
--- a/texk/web2c/pdftexdir/utils.c
+++ b/texk/web2c/pdftexdir/utils.c
@@ -33,7 +33,6 @@ with this program.  If not, see <http://www.gnu.org/licenses/>.
 #include "ptexlib.h"
 #include <png.h>
 #ifdef POPPLER_VERSION
-#include <poppler-config.h>
 #define xpdfVersion POPPLER_VERSION
 #define xpdfString "poppler"
 #else
diff --git a/texk/web2c/xetexdir/XeTeX_ext.c b/texk/web2c/xetexdir/XeTeX_ext.c
index 4968ee41..0aee4ee3 100644
--- a/texk/web2c/xetexdir/XeTeX_ext.c
+++ b/texk/web2c/xetexdir/XeTeX_ext.c
@@ -38,7 +38,10 @@ authorization from the copyright holders.
 
 #include <w2c/config.h>
 
+#ifndef POPPLER_VERSION
 #include <poppler-config.h>
+#endif
+
 #include <png.h>
 #include <zlib.h>
 #include <graphite2/Font.h>
diff -ru texlive-source-orig/texk/web2c/pdftexdir/pdftoepdf.cc texlive-source/texk/web2c/pdftexdir/pdftoepdf.cc
--- texlive-source-orig/texk/web2c/pdftexdir/pdftoepdf.cc	2020-01-07 22:42:03.717667766 +0000
+++ texlive-source/texk/web2c/pdftexdir/pdftoepdf.cc	2020-01-07 22:47:42.818888622 +0000
@@ -723,7 +723,7 @@
 #endif
     // initialize
     if (!isInit) {
-        globalParams = new GlobalParams();
+        globalParams.reset(new GlobalParams());
         globalParams->setErrQuiet(false);
         isInit = true;
     }
@@ -1108,6 +1108,5 @@
             delete_document(p);
         }
         // see above for globalParams
-        delete globalParams;
     }
 }
diff -ru texlive-source-orig/texk/web2c/pdftexdir/pdftosrc.cc texlive-source/texk/web2c/pdftexdir/pdftosrc.cc
--- texlive-source-orig/texk/web2c/pdftexdir/pdftosrc.cc	2020-01-07 22:42:03.714334387 +0000
+++ texlive-source/texk/web2c/pdftexdir/pdftosrc.cc	2020-01-07 22:50:08.070708640 +0000
@@ -79,7 +79,7 @@
         exit(1);
     }
     fileName = new GString(argv[1]);
-    globalParams = new GlobalParams();
+    globalParams.reset(new GlobalParams());
     doc = new PDFDoc(fileName);
     if (!doc->isOk()) {
         fprintf(stderr, "Invalid PDF file\n");
@@ -100,7 +100,7 @@
     if (objnum == 0) {
         srcStream = catalogDict.dictLookup("SourceObject");
         static char const_SourceFile[] = "SourceFile";
-        if (!srcStream.isStream(const_SourceFile)) {
+        if (!srcStream.isDict(const_SourceFile)) {
             fprintf(stderr, "No SourceObject found\n");
             exit(1);
         }
@@ -202,5 +202,4 @@
         fprintf(stderr, "Cross-reference table extracted to %s\n", outname);
     fclose(outfile);
     delete doc;
-    delete globalParams;
 }
--- texlive-source-build-svn51263-orig/texk/web2c/pdftexdir/pdftoepdf-poppler0.75.0.cc	2020-03-02 16:53:56.809524122 +0100
+++ texlive-source-build-svn51263/texk/web2c/pdftexdir/pdftoepdf-poppler0.75.0.cc	2020-03-02 16:59:18.494694749 +0100
@@ -757,7 +757,7 @@
     if (page_name) {
         // get page by name
         GString name(page_name);
-        LinkDest *link = pdf_doc->doc->findDest(&name);
+        const std::unique_ptr<LinkDest> link = pdf_doc->doc->findDest(&name);
         if (link == 0 || !link->isOk())
             pdftex_fail("PDF inclusion: invalid destination <%s>", page_name);
         Ref ref = link->getPageRef();
@@ -765,7 +765,6 @@
         if (page_num == 0)
             pdftex_fail("PDF inclusion: destination is not a page <%s>",
                         page_name);
-        delete link;
     } else {
         // get page by number
         if (page_num <= 0 || page_num > epdf_num_pages)
--- texlive-source-build-svn51263-orig/texk/web2c/pdftexdir/pdftoepdf.cc	2020-03-02 17:02:52.380146406 +0100
+++ texlive-source-build-svn51263/texk/web2c/pdftexdir/pdftoepdf.cc	2020-03-02 17:13:00.720178912 +0100
@@ -757,7 +757,7 @@
     if (page_name) {
         // get page by name
         GString name(page_name);
-        LinkDest *link = pdf_doc->doc->findDest(&name);
+        const std::unique_ptr<LinkDest> link = pdf_doc->doc->findDest(&name);
         if (link == 0 || !link->isOk())
             pdftex_fail("PDF inclusion: invalid destination <%s>", page_name);
         Ref ref = link->getPageRef();
@@ -765,7 +765,6 @@
         if (page_num == 0)
             pdftex_fail("PDF inclusion: destination is not a page <%s>",
                         page_name);
-        delete link;
     } else {
         // get page by number
         if (page_num <= 0 || page_num > epdf_num_pages)
